### PR TITLE
kodi: add module

### PR DESCRIPTION
**Pull request by dwagenk**
Kodi is a media center software.

### Description

This adds a module for configuring the kodi media center software via home manager. Since the settings are spread over multiple xml files and kodi is highly customizable through plugins I chose to implement configuration file generators for three different kinds of settings:

- `settings` allow modification of basically all builtin configuration options
- `sources` allow specifying the media sources for kodi
- `addonSettings` allow specifying settings for arbitrary addons

This allows modifying almost all kodi settings without having to write xml manually. I left out e.g. `favourites.xml` and `rssfeeds.xml` which again have slightly different requirenments on the xml structure.

This is my first ever module in nix, so please suggest improvements! Especially the option types for `settings` and `sources` (currently: `attrsOf(anything)`) could probably be improved, but I don't know how to specify "nested attribute sets and lists with the innermost beeing an `attrsOf str`".

### Checklist

- [x] Change is backwards compatible.

- [x] Code formatted with `./format`.

- [x] Code tested through `nix-shell --pure tests -A run.all`.
EDIT: Reran all tests after rebasing from `release-21.05` to `master` branch. 

- [x] Test cases updated/added. See [example](https://github.com/nix-community/home-manager/commit/f3fbb50b68df20da47f9b0def5607857fcc0d021#diff-b61a6d542f9036550ba9c401c80f00ef).

- [x] Commit messages are formatted like

    ```
    {component}: {description}

    {long description}
    ```

    See [CONTRIBUTING](https://github.com/nix-community/home-manager/blob/master/docs/contributing.adoc#sec-commit-style) for more information and [recent commit messages](https://github.com/nix-community/home-manager/commits/master) for examples.

- If this PR adds a new module

  - [x] Added myself as module maintainer. See [example](https://github.com/nix-community/home-manager/blob/068ff76a10e95820f886ac46957edcff4e44621d/modules/programs/lesspipe.nix#L6).

  - [x] Added myself and the module files to `.github/CODEOWNERS`.

### CODE REVIEW

**Comment by DWagenk** [comment]
Just pushed a changes to fixup:
- the package was missing in `modules/modules.nix`. Seems like I messed this up when polishing the commit before the first push.
- the documentation for `kodi.datadir` option had quotation errors. Aligned them to the `gpg.homedir` option documentation.

File: modules/programs/kodi.nix [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+    pkgs.stdenv.mkDerivation {
+      inherit name stylesheet;
+      nativeBuildInputs = [ pkgs.libxslt ];
+      phases = [ "installPhase" ];
+      installPhase = let xml = (builtins.toXML attrs);
+      in ''
+        xsltproc ${stylesheet} - <<EOF > $out
+        ${xml}
+        EOF
+      '';
```
Comment (rycee) [inline modules/programs/kodi.nix:92]: I think it probably is a bit clearer to simply inline the `xml` variable:
```suggestion
      installPhase = ''
        xsltproc ${stylesheet} - <<EOF > $out
        ${builtins.toXML attrs}
        EOF
      '';
```

File: modules/programs/kodi.nix [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+
+  attrsetToAdvancedSettingsXml = attrs: name:
+    let
+      stylesheet = builtins.toFile "stylesheet.xsl" ''
+        ${stylesheetCommonHeader}
+        ${stylesheetAdvancedSettingsRootTag}
+        ${stylesheetNestedTags}
+        ${stylesheetCommonFooter}
+      '';
+    in (attrsetToXml attrs name stylesheet);
```
Comment (rycee) [inline modules/programs/kodi.nix:103]: Can remove the unnecessary parentheses:
```suggestion
    in attrsetToXml attrs name stylesheet;
```

Same for the other instances below.

File: modules/programs/kodi.nix [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+      inherit name stylesheet;
+      nativeBuildInputs = [ pkgs.libxslt ];
+      phases = [ "installPhase" ];
+      installPhase = let xml = (builtins.toXML attrs);
+      in ''
+        xsltproc ${stylesheet} - <<EOF > $out
+        ${xml}
+        EOF
+      '';
+    };
```
Comment (rycee) [inline modules/programs/kodi.nix:93]: Might be a bit simpler with `runCommand` and `passAsFile` to be safe. Something like

```suggestion
    pkgs.runCommand name {
      nativeBuildInputs = [ pkgs.libxslt ];
      xml = builtins.toXML attrs;
      passAsFile = [ "xml" ];
    } ''
      xsltproc ${stylesheet} - < "$xmlPath" > "$out"
    ''
```

perhaps.

File: modules/programs/kodi.nix [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+in {
+  meta.maintainers = [ maintainers.dwagenk ];
+
+  options.programs.kodi = {
+    enable = mkEnableOption "Kodi";
+
+    package = mkOption {
+      type = types.package;
+      default = pkgs.kodi;
+      defaultText = literalExample "pkgs.kodi";
```
Comment (rycee) [inline modules/programs/kodi.nix:134]:
```suggestion
      defaultText = literalExpression "pkgs.kodi";
```

same elsewhere.

File: modules/programs/kodi.nix [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+        Can be used to specify extensions.
+      '';
+    };
+
+    datadir = mkOption {
+      type = types.path;
+      default = "${config.home.homeDirectory}/.kodi";
+      defaultText = literalExpression "\"\${config.home.homeDirectory}/.kodi\"";
+      description = "Directory to store configuration and metadata.";
+      example = literalExpression "\"\${config.xdg.dataHome}/kodi\"";
```
Comment (rycee) [inline modules/programs/kodi.nix:149]: Best to be consistent: type, default, example, and description last.

```suggestion
      example = literalExpression "\"\${config.xdg.dataHome}/kodi\"";
      description = "Directory to store configuration and metadata.";
```

File: modules/programs/kodi.nix [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+    };
+
+    settings = mkOption {
+      type = with types; nullOr (attrsOf anything);
+      default = null;
+      description = ''
+        Configuration to write to the <literal>advancedsettings.xml</literal>
+        file in kodis userdata directory. Settings specified here will be
+        immutable from inside kodi and be hidden from the GUI settings dialog.
+
```
Comment (rycee) [inline modules/programs/kodi.nix:159]: Need
```suggestion
        </para><para>
```
to indicate a paragraph.

File: modules/programs/kodi.nix [hunk 7]
```diff
@@ -0,0 +1,238 @@
...
+    };
+  };
+
+  config = mkIf cfg.enable (mkMerge [
+    {
+      home.packages = [ cfg.package ];
+      home.sessionVariables = { KODI_DATA = cfg.datadir; };
+    }
+
+    (mkIf (!isNull cfg.settings) {
```
Comment (rycee) [inline modules/programs/kodi.nix:220]:
```suggestion
    (mkIf (cfg.settings != null) {
```

same below

File: modules/programs/kodi.nix [hunk 8]
```diff
@@ -0,0 +1,238 @@
...
+      home.file."${cfg.datadir}/userdata/advancedsettings.xml".source =
+        (attrsetToAdvancedSettingsXml cfg.settings "kodi-advancedsettings.xml");
```
Comment (rycee) [inline modules/programs/kodi.nix:222]: Unnecessary parentheses
```suggestion
        attrsetToAdvancedSettingsXml cfg.settings "kodi-advancedsettings.xml";
```

Same below.

File: tests/modules/programs/kodi/example-addon-settings.nix [hunk 9]
```diff
@@ -0,0 +1,22 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+{
+  config = {
+    programs.kodi = {
+      enable = true;
+      package = pkgs.writeScriptBin "dummy-kodi" "";
```
Comment (rycee) [inline tests/modules/programs/kodi/example-addon-settings.nix:9]: There is now a special function to generate stub packages:

```suggestion
      package = config.lib.test.mkStubPackage { };;
```

File: modules/programs/kodi.nix [hunk 10]
```diff
@@ -0,0 +1,238 @@
...
+    addonSettings = mkOption {
+      type = with types; nullOr (attrsOf (attrsOf str));
+      default = null;
+      description = ''
+        Attribute set with the plugin namespace as toplevel key and the plugins
+        settings as lower level key/value pairs.
+
+        Kodi will still show the settings of plugins configured via this
+        mechanism in the GUI and they appear to be mutable. This however is
+        not the case and the settings will stay as specified via home-manager.
```
Comment (rycee) [inline modules/programs/kodi.nix:206]:
```suggestion
        not the case and the settings will stay as specified via Home Manager.
```

File: modules/programs/kodi.nix [hunk 11]
```diff
@@ -0,0 +1,238 @@
...
+      description = ''
+        Contents to populate the file <literal>sources.xml</literal> in kodis
+        userdata directory.
+
+        See <link xlink:href="https://kodi.wiki/view/Sources.xml"/> as
+        reference for how sources need to be specified.
+
+        Kodi will still show the dialogs to modify sources in the GUI and they
+        appear to be mutable. This however is not the case and the sources will
+        stay as specified via home-manager.
```
Comment (rycee) [inline modules/programs/kodi.nix:182]:
```suggestion
        stay as specified via Home Manager.
```

File: modules/programs/kodi.nix [hunk 12]
```diff
@@ -0,0 +1,238 @@
...
+    datadir = mkOption {
+      type = types.path;
+      default = "${config.home.homeDirectory}/.kodi";
+      defaultText = literalExpression "\"\${config.home.homeDirectory}/.kodi\"";
+      description = "Directory to store configuration and metadata.";
+      example = literalExpression "\"\${config.xdg.dataHome}/kodi\"";
+    };
+
+    settings = mkOption {
+      type = with types; nullOr (attrsOf anything);
```
Comment (rycee) [inline modules/programs/kodi.nix:153]: I think something like
```suggestion
      type = with types;
        let valueType = either str (attrsOf valueType) // {
              description = "attribute sets of strings";
            };
        in nullOr valueType;
```
may do what you want.

**Comment by rycee** [comment]
Thanks for the contribution! I've added some comments 🙂

File: modules/programs/kodi.nix (hunk 3 above)
Comment (DWagenk) [reply modules/programs/kodi.nix]: nice!

File: modules/programs/kodi.nix (hunk 12 above)
Comment (DWagenk) [reply modules/programs/kodi.nix]: It does, thanks for the suggestion!
For `sources` I also needed to add `listOf` to `valueType`.

**Comment by DWagenk** [comment]
I addressed your comments, @rycee. Thanks for the very helpful review!

**Comment by DWagenk** [comment]
Found a cross compilation error with libxslt. Will prepare a fix and add it to this PR in the next couple of days.

**Comment by DWagenk** [comment]
> Found a cross compilation error with libxslt. Will prepare a fix and add it to this PR in the next couple of days.

Done now. Please trigger the CI and merge.
Thanks!

**Comment by DWagenk** [comment]
Another fixup due to my messed up order of `<para></para>` in the option descriptions.

I inspected the failed CI build to run as much as possible locally:
- [x] check for `literalExample`
- [x] run `./format -c`
- [x] run `nix-shell . -A install`
- [x] run `nix-shell --arg enableBig false --pure tests -A run.all`

**Comment by DWagenk** [comment]
Should I do a rebase to current master? The contributing docs mention to start of the most recent `master` branch, but not whether rebasing should be done while the PR is still open. Local rebase to current `master` and rerunning the checks went without problems.

**Comment by berbiche** [comment]
> Whether rebasing should be done while the PR is still open

I would answer "yes" but it depends how old/stale the branch that is being merged is.
There has been no breaking changes recently so it does not really matter (imo).

**Comment by DWagenk** [comment]
Rebased to current master due to github showing merge conflicts. Please trigger the workflow and merge.

**Comment by rycee** [comment]
Thanks! I've rebased again and merged to master.

**Pull request merged**